File: PaletteViewController/PalletteTabVerticalTextLabel.m
//
//  MNVerticalTextLabel.m
//  TuiFramework
//
//  Created by Masatoshi Nishikata on 16/10/12.
//
//

#import "PalletteTabVerticalTextLabel.h"
#import <UIKit/UIKit.h>

@implementation PalletteTabVerticalTextLabel

@synthesize selected, highlighted;
@synthesize attributedString = attributedString_;
@synthesize highlightedAttributedString = highlightedAttributedString_;
@synthesize contentWidth;
@synthesize onRight;
@synthesize inverted;
@synthesize alignCenter = alignCenter_;
@synthesize alignCenterInsideFrame = alignCenterInsideFrame_;
@synthesize hasShadow = hasShadow_;
@synthesize shadowOffset = shadowOffset_;
@synthesize shadowBlur = shadowBlur_;
@synthesize shadowColor = shadowColor_;

#define MARGIN 7

- (id)initWithFrame:(CGRect)frame {
    
    self = [super initWithFrame:frame];
    if (self) {
        hasShadow_ = NO;
        shadowOffset_ = CGSizeMake(0,1);
        shadowBlur_ = 3.0;
        alignCenter_ = NO;
        alignCenterInsideFrame_ = NO;
        self.contentWidth = frame.size.width;
        self.backgroundColor = [UIColor clearColor];
        self.contentMode = UIViewContentModeRedraw;
        
        
    }
    return self;
}

-(void)setFrame:(CGRect)frame
{
    self.contentWidth = frame.size.width;
    [super setFrame:frame];
  [self setNeedsDisplay];

}

-(CTLineRef)line
{
    return line_;
}

-(void)setAttributedString:(NSAttributedString*)attributedString
{
    if( attributedString_ != attributedString )
    {
        attributedString_ = attributedString;
    }
    
    if( line_ )
    {
        CFRelease(line_);
        line_ = nil;
    }
    
    
    [self setNeedsDisplay];
}


//-(void)removeFromSuperview
//{
//
//    self.action = nil;
//
//
//    [super removeFromSuperview];
//}




-(void)dealloc
{
    attributedString_ = nil;
    
    if( line_ ) CFRelease(line_);
    
    //    self.action = nil;
    
    shadowColor_ = nil;
}

-(CGFloat)width
{
    
    if( !line_  )
    {
        line_ = CTLineCreateWithAttributedString((CFAttributedStringRef)attributedString_);
    }
    
    CGFloat width =  CTLineGetTypographicBounds(line_, nil, nil, nil);
    
    if( inverted ) width += 20;
    
    
    return width;
}

-(CGRect)textRect
{
    
    if( !line_  && attributedString_ )
    {
        line_ = CTLineCreateWithAttributedString((CFAttributedStringRef)attributedString_);
    }
    
    CGRect rect = self.bounds;
    CGRect contentRect;
    CGFloat descent;
    CGFloat ascent;
    CGFloat leading;
    CTLineGetTypographicBounds(line_, &ascent, &descent, &leading);
    
    CGFloat ty = roundf( (rect.size.height - descent - ascent)/2 +ascent ) +2;
    
    CGFloat width = CTLineGetTypographicBounds(line_, &ascent, &descent, &leading);
    CGFloat height = ascent;
    
    height += 10;
    contentRect = CGRectMake( (rect.size.width - width)/2, ty - height + 5, width, height);
    contentRect = CGRectIntegral(contentRect);
    
    return contentRect;
}



-(void)drawRect:(CGRect)rect
{
  if( !line_  && attributedString_ )
  {
    NSMutableAttributedString* mattr = (NSMutableAttributedString*)[attributedString_ mutableCopy];
    line_ = CTLineCreateWithAttributedString((CFAttributedStringRef)mattr);
    CGFloat width = CTLineGetTypographicBounds(line_, nil, nil, nil);
    
    if( rect.size.height < width + MARGIN && mattr.length > 2 ) {
      NSMutableDictionary *attributes = [[mattr attributesAtIndex:0 effectiveRange:nil] mutableCopy];
      UIFont* font = [attributes objectForKey: NSFontAttributeName];
      font = [UIFont fontWithName:font.fontName size:font.pointSize-2];
      
      attributes[NSFontAttributeName] = font;
      attributes[NSKernAttributeName] = @(-1);

      [mattr addAttributes:attributes range:NSMakeRange(0, mattr.length)];
      CFRelease(line_);
      line_ = CTLineCreateWithAttributedString((CFAttributedStringRef)mattr);
      width = CTLineGetTypographicBounds(line_, nil, nil, nil);
    }
    
    while( rect.size.height < width + MARGIN && mattr.length > 5 ) {
      NSDictionary *attributes = [mattr attributesAtIndex:mattr.length - 2 effectiveRange:nil];
      [mattr replaceCharactersInRange:NSMakeRange(mattr.length-4, 2) withString:@"︙"];
      [mattr addAttributes:attributes range:NSMakeRange(mattr.length - 2, 1)];
      CFRelease(line_);
      line_ = CTLineCreateWithAttributedString((CFAttributedStringRef)mattr);
      width = CTLineGetTypographicBounds(line_, nil, nil, nil);
    }
  }
  
  if( !attributedString_ || !line_ || attributedString_.length == 0 )
  {
    [super drawRect:rect];
    return;
  }
  
  CGFloat descent;
  CGFloat ascent;
  CGFloat leading;
  double width;
  
  width = CTLineGetTypographicBounds(line_, &ascent, &descent, &leading);
  
  CGContextRef ctx = UIGraphicsGetCurrentContext();
  CGContextSaveGState(ctx);
  
  CGAffineTransform transform = CGAffineTransformMakeScale(1, -1);
  transform = CGAffineTransformTranslate(transform, 0,0);
  CGContextSetTextMatrix(ctx, transform);
  
  CGAffineTransform t;
  
  // Drawing left part
  if( self.alignedToLeft ) {
    t = CGAffineTransformMakeTranslation( rect.size.width - 22, MARGIN );

  }else {
    t = CGAffineTransformMakeTranslation( 12, MARGIN );
//    t = CGAffineTransformMakeTranslation( rect.size.width/2-7, MARGIN );
  }
  CGContextConcatCTM(ctx, t);
  
  NSArray* runs = (NSArray*)CTLineGetGlyphRuns(line_);
  for( id run in runs )
  {
    CTRunDraw((CTRunRef)run, ctx, CFRangeMake(0, 0));
  }
  
  CGContextRestoreGState(ctx);
}

@end

File: PaletteViewController/PalletteTabVerticalTextLabel.swift
//
//  PalletteTabVerticalTextLabel.swift
//  PaletteViewController
//
//  Created by Masatoshi Nishikata on 14/08/23.
//  Copyright © 2023 Catalystwo Limited. All rights reserved.
//

import Foundation
import UIKit

class PalletteTabVerticalTextLabel: UIView {
  var selected: Bool = false
  var highlighted: Bool = false
  var attributedString: NSAttributedString? {
    didSet {
      if line != nil {
        line = nil
      }
      setNeedsDisplay()
    }
  }
  var MARGIN = 7.0
  var highlightedAttributedString: NSAttributedString?
  var contentWidth: CGFloat
  var onRight: Bool
  var inverted: Bool
  var alignCenter: Bool
  var alignCenterInsideFrame: Bool
  var hasShadow: Bool
  var shadowOffset: CGSize
  var shadowBlur: CGFloat
  var shadowColor: UIColor?
  
  private var line: CTLine?
  
  override init(frame: CGRect) {
    contentWidth = frame.size.width
    onRight = false
    inverted = false
    alignCenter = false
    alignCenterInsideFrame = false
    hasShadow = false
    shadowOffset = CGSize(width: 0, height: 1)
    shadowBlur = 3.0
    shadowColor = nil
    
    super.init(frame: frame)
    
    backgroundColor = .clear
    contentMode = .redraw
  }
  
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  override var frame: CGRect {
    didSet {
      contentWidth = frame.size.width
      setNeedsDisplay()
    }
  }
  
  var lineRef: CTLine? {
    return line
  }
  
  func width() -> CGFloat {
    if line == nil, let attributedString = attributedString {
      line = CTLineCreateWithAttributedString(attributedString as CFAttributedString)
    }
    
    var width = CTLineGetTypographicBounds(line!, nil, nil, nil)
    if inverted {
      width += 20
    }
    
    return width
  }
  
  func textRect() -> CGRect {
    if line == nil, let attributedString = attributedString {
      line = CTLineCreateWithAttributedString(attributedString as CFAttributedString)
    }
    
    var rect = bounds
    var ascent: CGFloat = 0
    var descent: CGFloat = 0
    var leading: CGFloat = 0
    CTLineGetTypographicBounds(line!, &ascent, &descent, &leading)
    
    let ty = round((rect.size.height - descent - ascent) / 2 + ascent) + 2
    var width: CGFloat = 0
    var height = ascent
    height += 10
    
    CTLineGetTypographicBounds(line!, &ascent, &descent, &leading)
    width = CTLineGetTypographicBounds(line!, &ascent, &descent, &leading)
    var contentRect = CGRect(x: (rect.size.width - width) / 2, y: ty - height + 5, width: width, height: height)
    contentRect = contentRect.integral
    
    return contentRect
  }
  
  override func draw(_ rect: CGRect) {
    if line == nil, let attributedString = attributedString {
      let mattr = NSMutableAttributedString(attributedString: attributedString)
      line = CTLineCreateWithAttributedString(mattr as CFAttributedString)
      var width = CTLineGetTypographicBounds(line!, nil, nil, nil)
      
      if rect.size.height < width + MARGIN && mattr.length > 2 {
        var attributes = mattr.attributes(at: 0, effectiveRange: nil)
        if let font = attributes[NSAttributedString.Key.font] as? UIFont {
          let newFont = UIFont(name: font.fontName, size: font.pointSize - 2)
          attributes[NSAttributedString.Key.font] = newFont
          attributes[NSAttributedString.Key.kern] = -1
          mattr.addAttributes(attributes, range: NSRange(location: 0, length: mattr.length))
          line = nil
          line = CTLineCreateWithAttributedString(mattr as CFAttributedString)
          width = CTLineGetTypographicBounds(line!, nil, nil, nil)
        }
      }
      
      while rect.size.height < width + MARGIN && mattr.length > 5 {
        let attributes = mattr.attributes(at: mattr.length - 2, effectiveRange: nil)
        mattr.replaceCharacters(in: NSRange(location: mattr.length - 4, length: 2), with: "︙")
        mattr.addAttributes(attributes, range: NSRange(location: mattr.length - 2, length: 1))
        line = nil
        line = CTLineCreateWithAttributedString(mattr as CFAttributedString)
        width = CTLineGetTypographicBounds(line!, nil, nil, nil)
      }
    }
    
    guard let attributedString = attributedString, let line = line, attributedString.length > 0 else {
      super.draw(rect)
      return
    }
    
    var ascent: CGFloat = 0
    var descent: CGFloat = 0
    var leading: CGFloat = 0
    var width: Double = 0
    
    width = CTLineGetTypographicBounds(line, &ascent, &descent, &leading)
    
    guard let ctx = UIGraphicsGetCurrentContext() else {
      return
    }
    
    ctx.saveGState()
    
    var transform = CGAffineTransform(scaleX: 1, y: -1)
    transform = transform.translatedBy(x: 0, y: 0)
    ctx.textMatrix = transform
    
    var t = CGAffineTransform()
    
    if alignCenter {
      t = CGAffineTransform(translationX: rect.size.width - 22, y: MARGIN)
    } else {
      t = CGAffineTransform(translationX: 12, y: MARGIN)
    }
    
    ctx.concatenate(t)
    
    let runs = CTLineGetGlyphRuns(line) as! [CTRun]
    for run in runs {
      CTRunDraw(run, ctx, CFRangeMake(0, 0))
    }
    
    ctx.restoreGState()
  }
}
